File: src/main/java/org/cubexmc/metro/util/BedrockPlayerUtil.java
package org.cubexmc.metro.util;

import java.lang.reflect.Method;
import java.util.UUID;

import org.bukkit.entity.Player;

/**
 * Optional Bedrock-player detection through Geyser/Floodgate without a hard API dependency.
 */
public final class BedrockPlayerUtil {

    private BedrockPlayerUtil() {
    }

    public static boolean isBedrockPlayer(Player player) {
        if (player == null) {
            return false;
        }
        UUID uuid = player.getUniqueId();
        return isGeyserPlayer(uuid) || isFloodgatePlayer(uuid);
    }

    private static boolean isGeyserPlayer(UUID uuid) {
        try {
            Class<?> apiClass = Class.forName("org.geysermc.geyser.api.GeyserApi");
            Method apiMethod = apiClass.getMethod("api");
            Object api = apiMethod.invoke(null);
            if (api == null) {
                return false;
            }
            Method isBedrockPlayer = apiClass.getMethod("isBedrockPlayer", UUID.class);
            return Boolean.TRUE.equals(isBedrockPlayer.invoke(api, uuid));
        } catch (ReflectiveOperationException | LinkageError ignored) {
            return false;
        }
    }

    private static boolean isFloodgatePlayer(UUID uuid) {
        try {
            Class<?> apiClass = Class.forName("org.geysermc.floodgate.api.FloodgateApi");
            Method getInstance = apiClass.getMethod("getInstance");
            Object api = getInstance.invoke(null);
            if (api == null) {
                return false;
            }
            Method isFloodgatePlayer = apiClass.getMethod("isFloodgatePlayer", UUID.class);
            return Boolean.TRUE.equals(isFloodgatePlayer.invoke(api, uuid));
        } catch (ReflectiveOperationException | LinkageError ignored) {
            return false;
        }
    }
}

File: src/main/java/org/cubexmc/metro/util/BedrockPlayerUtil.kt
package org.cubexmc.metro.util

import org.bukkit.entity.Player
import java.util.UUID

/**
 * Optional Bedrock-player detection through Geyser/Floodgate without a hard API dependency.
 */
object BedrockPlayerUtil {

    @JvmStatic
    fun isBedrockPlayer(player: Player?): Boolean {
        if (player == null) {
            return false
        }
        // 测试里的 mock Player 会返回 null uuid，保持 Java 版本的宽容语义
        val uuid: UUID? = player.uniqueId
        return isGeyserPlayer(uuid) || isFloodgatePlayer(uuid)
    }

    private fun isGeyserPlayer(uuid: UUID?): Boolean =
        try {
            val apiClass = Class.forName("org.geysermc.geyser.api.GeyserApi")
            val api = apiClass.getMethod("api").invoke(null)
            if (api == null) {
                false
            } else {
                val isBedrockPlayer = apiClass.getMethod("isBedrockPlayer", UUID::class.java)
                java.lang.Boolean.TRUE == isBedrockPlayer.invoke(api, uuid)
            }
        } catch (_: ReflectiveOperationException) {
            false
        } catch (_: LinkageError) {
            false
        }

    private fun isFloodgatePlayer(uuid: UUID?): Boolean =
        try {
            val apiClass = Class.forName("org.geysermc.floodgate.api.FloodgateApi")
            val api = apiClass.getMethod("getInstance").invoke(null)
            if (api == null) {
                false
            } else {
                val isFloodgatePlayer = apiClass.getMethod("isFloodgatePlayer", UUID::class.java)
                java.lang.Boolean.TRUE == isFloodgatePlayer.invoke(api, uuid)
            }
        } catch (_: ReflectiveOperationException) {
            false
        } catch (_: LinkageError) {
            false
        }
}
